### PR TITLE
Feature/special price

### PR DESCRIPTION
**Pull request by xandrej**
__Tasks:__
- [x] done
- [ ] todo
- [ ] got feedback

__Informations: Backend only

| Q            | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Tests pass?  | yes
| Fixed tickets | https://github.com/massiveart/POOL-ALPIN/issues/1149
| Doc          | none

### CODE REVIEW

File: Api/Product.php [hunk 1]
```diff
@@ -916,4 +915,53 @@ public function setAreGrossPrices($areGrossPrices = false)
     {
         $this->entity->setAreGrossPrices($areGrossPrices);
     }
+
+    /**
+     * getSupplier
```
Comment (notsure) [inline Api/Product.php]: first letter in php docs is Upper case

File: Api/Product.php [hunk 2]
```diff
@@ -916,4 +915,53 @@ public function setAreGrossPrices($areGrossPrices = false)
...
+     * Has this name because SuluProduct provides a supplier which is just an array
```
Comment (notsure) [inline Api/Product.php]: add empty line

File: Api/Product.php [hunk 3]
```diff
@@ -916,4 +915,53 @@ public function setAreGrossPrices($areGrossPrices = false)
...
+     * @return Sulu\Bundle\ContactBundle\Entity\Account
```
Comment (notsure) [inline Api/Product.php]: @return ApiAccount|null

File: Api/Product.php [hunk 4]
```diff
@@ -916,4 +915,53 @@ public function setAreGrossPrices($areGrossPrices = false)
...
+     */
+    public function getApiSupplier()
+    {
+        $account = $this->entity->getSupplier();
+        if ($account) {
+            return new ApiAccount($account, $this->locale);
+        }
+        return null;
```
Comment (notsure) [inline Api/Product.php]: add blank before the return

File: Api/Product.php [hunk 5]
```diff
@@ -916,4 +915,53 @@ public function setAreGrossPrices($areGrossPrices = false)
...
+    }
+
+    /**
+     * Returns the special prices for the product
+     * @return \Sulu\Bundle\ProductBundle\Api\SpecialPrice[]
```
Comment (notsure) [inline Api/Product.php]: @return is always the last entry in php docs

File: Api/Product.php [hunk 6]
```diff
@@ -916,4 +915,53 @@ public function setAreGrossPrices($areGrossPrices = false)
...
+     */
+    public function getSpecialPrices()
+    {
+        $specialPrices = $this->entity->getSpecialPrices();
+
+        $specialPricesList = array();
+        foreach ($specialPrices as $specialPrice) {
+            $specialPricesList[] = new SpecialPrice($specialPrice, $this->locale);
+        }
+        return $specialPricesList;
```
Comment (notsure) [inline Api/Product.php]: blank line

File: Api/Product.php [hunk 7]
```diff
@@ -916,4 +915,53 @@ public function setAreGrossPrices($areGrossPrices = false)
...
+    }
+
+    /**
+     * Adds a special price to the product
+     * @param SpecialPriceEntity $specialPrice
```
Comment (notsure) [inline Api/Product.php]: php doc format

File: Api/SpecialPrice.php [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        $this->entity = $entity;
+        $this->locale = $locale;
+    }
+
+    /**
+     * Returns the id of the Special price
+     *
+     * @VirtualProperty
+     * @SerializedName("id")
+     * @return int
```
Comment (notsure) [inline Api/SpecialPrice.php]: blank before return; 
@return integer

File: Api/SpecialPrice.php [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    {
+        return $this->entity->getId();
+    }
+
+    /**
+     * Get price
+     *
+     * @VirtualProperty
+     * @SerializedName("price")
+     * @return float
```
Comment (notsure) [inline Api/SpecialPrice.php]: php doc syntax

File: Api/SpecialPrice.php [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+     */
+    public function getPrice()
+    {
+        return $this->entity->getPrice();
+    }
+
+    /**
+     * Get start date
+     * @VirtualProperty
```
Comment (notsure) [inline Api/SpecialPrice.php]: php doc syntax

File: Api/SpecialPrice.php [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    {
+        return $this->entity->getStart();
+    }
+
+    /**
+     * Get end date
+     *
+     * @VirtualProperty
+     * @SerializedName("End")
+     * @return \DateTime
```
Comment (notsure) [inline Api/SpecialPrice.php]: php doc syntax

File: Api/SpecialPrice.php [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    {
+        return $this->entity->getEnd();
+    }
+
+    /**
+     * Get currency
+     *
+     * @VirtualProperty
+     * @SerializedName("Currency")
+     * @return \Sulu\Bundle\ProductBundle\Entity\Currency
```
Comment (notsure) [inline Api/SpecialPrice.php]: php doc syntax

File: Api/Product.php [hunk 13]
```diff
@@ -916,4 +915,53 @@ public function setAreGrossPrices($areGrossPrices = false)
...
+     */
+    public function addSpecialPrice(SpecialPriceEntity $specialPrice)
+    {
+        $this->entity->addSpecialPrice($specialPrice);
+    }
+
+    /**
+     * Removes a special price to the product
+     * @param SpecialPriceEntity $specialPrice
```
Comment (christianbader) [inline Api/Product.php]: one blank line after headline in php doc

File: Api/Product.php [hunk 14]
```diff
@@ -916,4 +915,53 @@ public function setAreGrossPrices($areGrossPrices = false)
...
+     * Adds a special price to the product
+     * @param SpecialPriceEntity $specialPrice
+     */
+    public function addSpecialPrice(SpecialPriceEntity $specialPrice)
+    {
+        $this->entity->addSpecialPrice($specialPrice);
+    }
+
+    /**
+     * Removes a special price to the product
```
Comment (christianbader) [inline Api/Product.php]: ... from the product?

File: Api/SpecialPrice.php [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+    /**
+     * Get product
+     *
+     * @return \Sulu\Bundle\ProductBundle\Entity\ProductInterface
+     */
+    public function getProduct()
+    {
+        return $this->entity->getProduct();
+    }
+
```
Comment (christianbader) [inline Api/SpecialPrice.php]: remove empty line

File: Api/Product.php (hunk 1 above)
Comment (turbo-ele) [inline Api/Product.php]: remove this comment

File: Api/SpecialPrice.php [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+namespace Sulu\Bundle\ProductBundle\Api;
+
+use Sulu\Component\Rest\ApiWrapper;
+use Sulu\Bundle\ProductBundle\Entity\SpecialPrice as SpecialPriceEntity;
+use JMS\Serializer\Annotation\VirtualProperty;
+use JMS\Serializer\Annotation\SerializedName;
+use JMS\Serializer\Annotation\ExclusionPolicy;
+
+/**
+ * The Status class which will be exported to the API
```
Comment (turbo-ele) [inline Api/SpecialPrice.php]: change description (it'S not a status class)

File: Product/ProductManager.php [hunk 17]
```diff
@@ -902,6 +903,69 @@ public function save(
...
+                    $specialPrice->setPrice($specialPriceData['price']);
+
+                    $startDate = date_create($specialPriceData['start']);
+                    $specialPrice->setStart($startDate);
+
+                    $endDate = date_create($specialPriceData['end']);
+                    $specialPrice->setEnd($endDate);
+                }
+            }
+
```
Comment (turbo-ele) [inline Product/ProductManager.php]: remove blank line

File: Product/ProductManager.php [hunk 18]
```diff
@@ -902,6 +903,69 @@ public function save(
...
+            //array for local special prices storage
+            $specialPrices = array();
+
+            // array of currency codes to be used as keys for special prices
+            $currencyCodes = array();
+
+            // create array of special price currency codes in json request
+            foreach($specialPricesData as $specialPrice) {
+                array_push($currencyCodes, $specialPrice['currency']['code']);
+            }
```
Comment (turbo-ele) [inline Product/ProductManager.php]: add array_unique to unify currency codes (just in case)

File: Product/ProductManager.php [hunk 19]
```diff
@@ -902,6 +903,69 @@ public function save(
...
+
+                    $product->addSpecialPrice($specialPrice);
+                    $this->em->persist($specialPrice);
+                // else update the already existing special price
+                } else {
+                    $specialPrice = $specialPrices[$specialPriceData['currency']['code']]->getEntity();
+
+                    $specialPrice->setPrice($specialPriceData['price']);
+
+                    $startDate = date_create($specialPriceData['start']);
```
Comment (turbo-ele) [inline Product/ProductManager.php]: another possibility would be to call

```{php}
$startDate = new \DateTime($specialPriceData['start']);
```

this is actually just an alias for that

File: Resources/config/doctrine/Product.orm.xml [hunk 20]
```diff
@@ -23,7 +23,14 @@
                 <cascade-persist/>
             </cascade>
         </one-to-many>
-
+        <one-to-many target-entity="Sulu\Bundle\ProductBundle\Entity\SpecialPrice"
+                     mapped-by="product"
+                     field="specialPrices">
```
Comment (turbo-ele) [inline Resources/config/doctrine/Product.orm.xml]: move field at first position (should be the first attribute)

File: Resources/config/doctrine/Product.orm.xml [hunk 21]
```diff
@@ -23,7 +23,14 @@
...
+            <cascade>
```
Comment (turbo-ele) [inline Resources/config/doctrine/Product.orm.xml]: imho: i would remove automatic cascading

File: Product/ProductManager.php (hunk 19 above)
Comment (chirimoya) [inline Product/ProductManager.php]: +1

**Comment by turbo-ele** [comment]
works only for one special-price per currency
make it fast forward
but ACK

**Pull request merged**